bluefog/torch/topology_util.py

```diff
@@ -0,0 +1,104 @@
+from typing import Any, List, Optional, Union
+import collections
+
+import numpy as np
+import torch
+import bluefog.torch as bf
+
+from bluefog.common.topology_util import (
+    GetRecvWeights,
+    GetSendWeights,
+    IsRegularGraph,
+    IsTopologyEquivalent,
+)
+
+from bluefog.common.topology_util import (
+    ExponentialTwoGraph,
+    ExponentialGraph,
+    FullyConnectedGraph,
+    MeshGrid2DGraph,
+    RingGraph,
+    StarGraph,
+    SymmetricExponentialGraph,
+)
+
+from bluefog.common.topology_util import (
+    GetDynamicOnePeerSendRecvRanks,
+    GetExp2DynamicSendRecvMachineRanks,
+    GetInnerOuterRingDynamicSendRecvRanks,
+    GetInnerOuterExpo2DynamicSendRecvRanks,
+)
+
+
+def _check_ranks(rank_list: List[Any], self_rank: int, size: int) -> [bool, str]:
+    for rank in rank_list:
+        if not isinstance(rank, int):
+            return False, "contain element that is not integer."
+        if (rank < 0) or (rank >= size):
+            return False, "contain element that is not between 0 and size-1."
+    if len(set(rank_list)) != len(rank_list):
+        return False, "contain duplicated elements."
+    if self_rank in rank_list:
+        return False, "contain self rank."
+    return True, ""
+
+
+def infer_destination_source_ranks(
+    *,
+    dst_ranks: Optional[List[int]] = None,
+    src_ranks: Optional[List[int]] = None,
+    construct_adjacency_matrix: bool = False,
+) -> Union[List[int], np.array]:
+    """Infer the destination(source) ranks from source(destination) ranks.
+
+    Args:
+        dst_ranks: A list of destination ranks. If provided the dst_link, a corresponding
+            src_ranks will be returned.
+        src_ranks: A list of destination ranks. If provided the src_ranks, a corresponding
+            dst_link will be returned.
+        construct_adjacency_matrix: If true, adjacency matrix will be return instead.
+            Element w_{ij} represents the weights sending from node i to node j. 
+            We use column normalized style, i.e. the sum of receiving weight is 1.
+
+    Raises:
+        ValueError: If dst_ranks or src_ranks does not contain integer from 0 to size-1.
```

Comment:
Mention dst_ranks and src_ranks should not be given at the same time

Reply:
Split the function into two